rp2: fix atomic section

dpgeorge opened the pull request:
This is an attempt to fix #12980 and #13288.

~~There are two fixes to the `MICROPY_BEGIN_ATOMIC_SECTION`/`
MICROPY_END_ATOMIC_SECTION` on the rp2 port:~~
- ~~Use a recursive mutex to prevent deadlocks on the same thread, in the case an IRQ races against the thread level (on the same core) when acquiring the atomic section (mutex may be taken but IRQs not yet disabled, then the IRQ waits forever trying to acquire the mutux).~~
- ~~Unlock the mutex if it was locked, not if `core1_entry` is still non-null, to fix the case where the second core finishes (and sets `core1_entry` to `NULL`) while the first core is in the middle of an atomic operation.~~

Edit: the two changes are now:
1. split out multicore lockout handling so it's only done when doing a flash erase/write
2. change the ATOMIC_SECTION macros so they use new mutex enter/exit functions that also disable/restore interrupts (atomically).

dpgeorge committed:
rp2/rp2_flash: Lockout second core only when doing flash erase/write.

Using the multicore lockout feature in the general atomic section makes it
much more difficult to get correct.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2/mpthreadport: Fix race with IRQ when entering atomic section.

Prior to this commit there is a potential deadlock in
mp_thread_begin_atomic_section(), when obtaining the atomic_mutex, in the
following situation:
- main thread calls mp_thread_begin_atomic_section() (for whatever reason,
  doesn't matter)
- the second core is running so the main thread grabs the mutex via the
  call mp_thread_mutex_lock(&atomic_mutex, 1), and this succeeds
- before the main thread has a chance to run save_and_disable_interrupts()
  a USB IRQ comes in and the main thread jumps off to process this IRQ
- that USB processing triggers a call to the dcd_event_handler() wrapper
  from commit bcbdee235719d459a4cd60d51021454fba54cd0f
- that then calls mp_sched_schedule_node()
- that then attempts to obtain the atomic section, calling
  mp_thread_begin_atomic_section()
- that call then blocks trying to obtain atomic_mutex
- core0 is now deadlocked on itself, because the main thread has the mutex
  but the IRQ handler (which preempted the main thread) is blocked waiting
  for the mutex, which will never be free

The solution in this commit is to use mutex enter/exit functions that also
atomically disable/restore interrupts.

Fixes issues #12980 and #13288.

Signed-off-by: Damien George <damien@micropython.org>

mendenm commented:
It works!  At least  the simple test  script runs both processors and reports results form them.

Excellent.

projectgus commented on ports/rp2/mpthreadport.c:
```diff
@@ -45,16 +45,15 @@ STATIC uint32_t *core1_stack = NULL;
-        // suspend the other core.
-        multicore_lockout_start_blocking();
+        recursive_mutex_enter_blocking(&atomic_mutex);
+        ++atomic_taken;
```
Doesn't `recursive_mutex_enter_blocking()` provide this recursion count function?

dpgeorge committed:
rp2/mutex_extra: Implement additional mutex functions.

These allow entering/exiting a mutex and also disabling/restoring
interrupts, in an atomic way.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I have changed this to use a new set of mutex functions that also disable/restore interrupts when obtaining/releasing the mutex.

projectgus commented:
New approach looks good to me!
